proposals/ref-readonly-parameters.md

```diff
@@ -0,0 +1,104 @@
+`in` parameters allow both lvalues and rvalues and can be used without any annotation at the callsite.
+However, APIs which capture or return references from their parameters would like to disallow rvalues and also enforce some indication at the callsite that a reference is being captured.
+`ref readonly` parameters are ideal in such cases as they allow only lvalues and warn if used without any annotation.
+
+Furthermore, there are APIs that need only read-only references but use
+- `ref` parameters since they were introduced before `in` became available and changing to `in` would be a source and binary breaking change, e.g., `QueryInterface`, or
+- `in` parameters to accept readonly references even though passing rvalues to them doesn't really make sense, e.g., `ReadOnlySpan<T>..ctor(in T value)`, or
+- `ref` parameters to disallow rvalues even though they don't mutate the passed reference, e.g., `Unsafe.IsNullRef`.
+
+These APIs could migrate to `ref readonly` parameters without breaking users.
+For details on binary compatibility, see the proposed [metadata encoding](#metadata).
+Specifically, changing
+- `ref` → `ref readonly` would only be a binary breaking change for virtual methods,
+- `ref` → `in` would also be a binary breaking change for virtual methods, but not a source breaking change (because the rules change to only warn for `ref` arguments passed to `in` parameters),
+- `in` → `ref readonly` would not be a breaking change (but no callsite annotation would result in a warning).
+
+In the opposite direction, changing
+- `ref readonly` → `ref` would be potentially a source breaking change (unless only `ref` callsite annotation was used), and a binary breaking change for virtual methods,
+- `ref readonly` → `in` would not be a breaking change (but `ref` callsite annotation would result in a warning).
+
+## Detailed design
+[design]: #detailed-design
+
+No changes in grammar are necessary.
+Specification will be extended to allow `ref readonly` modifiers for parameters with the same rules as specified for `in` parameters in [their proposal](https://github.com/dotnet/csharplang/blob/c8c1615fcad4ca016a97b7260ad497aad53ebc78/proposals/csharp-7.2/readonly-ref.md), except where explicitly changed in this proposal.
+
+### Metadata encoding
+[metadata]: #metadata-encoding
+
+As a reminder,
+- `ref` parameters are emitted as plain byref types (`T&` in IL),
+- `in` parameters are like `ref` plus they are annotated with `System.Runtime.CompilerServices.IsReadOnlyAttribute`.
+  In C# 7.3 and later, they are also emitted with `[in]` and if virtual, `modreq(System.Runtime.InteropServices.InAttribute)`.
+
+`ref readonly` parameters will be emitted as `[in] T&`, plus annotated with the following attribute:
+
+```cs
+namespace System.Runtime.CompilerServices
+{
+    [AttributeUsage(AttributeTargets.Parameter, Inherited = false)]
```

Comment:
I don't think we mentioned this, but should this also have `[EditorBrowsable(EditorBrowsableState.Never)]`? Eg. `IsReadOnlyAttribute` has it. It would seem the two would basically be equivalent in use? Ie. compiler only.

Reply:
Just checked and eg. `[RefSafetyRules]` doesn't have it either, so I take it it's not needed. Nvm 😄

Reply:
`IsReadOnlyAttribute` does have it:
<img width="997" alt="image" src="https://user-images.githubusercontent.com/25316035/236807674-7ae400ec-2e04-4e1b-b000-d9978b05b802.png">
<img width="563" alt="image" src="https://user-images.githubusercontent.com/25316035/236807692-11ca49c1-a053-4378-890b-a83a15e7dced.png">
<img width="689" alt="image" src="https://user-images.githubusercontent.com/25316035/236807752-792d60c2-1258-461c-a3f1-9f24e6b8b25e.png">

I don't see why it has it honestly, but if it has it, then `RequiresLocationAttribute` probably should too.

Reply:
Seems that other compiler-synthesized attributes like `ScopedRefAttribute` or `RefSafetyRulesAttribute` don't have it, not sure why `IsReadOnlyAttribute` has it, but it looks like an exception rather than the rule.